feat: Add collision completed listener (#2896)
flame-engine/flame: doc/flame/collision_detection.md

position is instead returned.

### Collision order

> If a `PositionComponent` collides with more than one other object within a given time step, then

Reviewer: ```suggestion
If a `Hitbox` collides with more than one other `Hitbox` within a given time step, then
```

the `onCollision` callbacks will be called in an essentially random order.  In some cases this can
be a problem, such as in a bouncing ball game where the trajectory of the ball can differ depending
on which other object was hit first.  To help resolve this the `collisionsCompletedNotifier`
listener can be used - this triggers at the end of the collision detection process.

An example of how this might be used is to add a local variable in your `PositionComponent` to save

> the other components with which this one is colliding:

Reviewer: ```suggestion
the other components with which it's colliding with:
```

`List<PositionComponent> collisionComponents = [];`.  The `onCollision` callback is then used to

> save all the other `PositionComponent`s to this list:

Reviewer: All of these sentences start with two spaces, there should only be one to be consistent with the rest of the docs.

Author: Sorry, that's a British English thing, I've changed that.

```dart
@override
void onCollision(Set<Vector2> intersectionPoints, PositionComponent other) {
  collisionComponents.add(other);
  super.onCollision(intersectionPoints, other);
}

```

> Finally, one adds a listener to the `onLoad` method of the `PositionComponent` to call a routine

Reviewer: ```suggestion
Finally, one adds a listener to the `onLoad` method of the `PositionComponent` to call a function
```

which will resolve how the collisions should be dealt with:

```dart
(game as HasCollisionDetection).collisionDetection.collisionsCompletedNotifier.
                                addListener(() {
  resolveCollisions();

> });

Reviewer: Formatting

Author: I have to break the line (as it's too long otherwise).  I've now done it in a way that looks nice to me, happy to hear your thoughts.

Reviewer: if you put it in an IDE and or https://dartpad.dev it will correctly format it for you, since it follows 80 width by default

Author: Couldn't get VScode, dartpad or tutorialspoint to do it automatically for me. :-(

position is instead returned.

### Collision order

If a `Hitbox` collides with more than one other `Hitbox` within a given time step, then
the `onCollision` callbacks will be called in an essentially random order. In some cases this can
be a problem, such as in a bouncing ball game where the trajectory of the ball can differ depending
on which other object was hit first. To help resolve this the `collisionsCompletedNotifier`
listener can be used - this triggers at the end of the collision detection process.

An example of how this might be used is to add a local variable in your `PositionComponent` to save
the other components with which it's colliding:
`List<PositionComponent> collisionComponents = [];`. The `onCollision` callback is then used to
save all the other `PositionComponent`s to this list:

```dart
@override
void onCollision(Set<Vector2> intersectionPoints, PositionComponent other) {
  collisionComponents.add(other);
  super.onCollision(intersectionPoints, other);
}

```

Finally, one adds a listener to the `onLoad` method of the `PositionComponent` to call a function
which will resolve how the collisions should be dealt with:

```dart

> (game as HasCollisionDetection).collisionDetection
>                                .collisionsCompletedNotifier
>                                .addListener(() {
>   resolveCollisions();
> });

Reviewer: ```suggestion
(game as HasCollisionDetection)
    .collisionDetection
    .collisionsCompletedNotifier
    .addListener(() {
  resolveCollisions();
});
```
This is the correct formatting
